s3/src/main/scala/tamer/s3/package.scala

```diff
@@ -60,124 +56,94 @@ package object s3 {
+      q: Queue[(K, V)]
+  ): ZIO[R with zio.s3.S3, TamerError, S] =
     (for {
-      log         <- logTask
-      nextInstant <- getNextInstant(keysR, afterwards, setup.prefix, setup.zonedDateTimeFormatter.value)
-      nextState = LastProcessedInstant(nextInstant.getOrElse(afterwards.instant))
-      _    <- log.debug(s"Next state computed to be $nextState")
-      keys <- keysR.get
-      _ <- whenSome(nextInstant) { instant =>
-        val optKey = deriveKey(instant, setup.zonedDateTimeFormatter.value, keys)
-        log.debug(s"Will ask for key $optKey") *> optKey
-          .map(key =>
-            zio.s3
-              .getObject(setup.bucketName, key)
-              .transduce(setup.transducer)
-              .foreach(value => q.offer(tamer.s3.S3Object(setup.bucketName, key) -> value))
-          )
-          .getOrElse(ZIO.fail(TamerError(s"File not found with date $instant")))
-      }
-    } yield nextState).mapError(e => TamerError("Error while doing iteration", e))
-
-  private final def dedupingIteration[R, V <: Product: Encoder: Decoder: SchemaFor](
-      setup: Setup[R, V],
-      keysR: KeysR
-  )(
-      currentState: LastProcessedInstant,
-      q: Queue[(tamer.s3.S3Object, V)]
-  ): ZIO[R with zio.s3.S3 with Clock, TamerError, LastProcessedInstant] =
-    (iteration(setup, keysR)(currentState, q) <&> logTask)
-      .flatMap { case (nextState, log) =>
-        if (nextState == currentState)
-          log.info(s"State is still $currentState, waiting ${setup.minimumIntervalForBucketFetch}") *>
-            clock.sleep(setup.minimumIntervalForBucketFetch) *>
-            dedupingIteration(setup, keysR)(currentState, q)
-        else UIO(nextState)
-      }
-      .mapError(t => TamerError("Error while wrapping iteration", t))
-
-  private final def dedupingIterationBlocking[R, V <: Product: Encoder: Decoder: SchemaFor](
-      setup: Setup[R, V],
-      keysR: KeysR
-  )(currentState: LastProcessedInstant, q: Queue[(tamer.s3.S3Object, V)]) =
-    blocking(dedupingIteration(setup, keysR)(currentState, q))
-  // FIXME: should spawn thread only if necessary but since for S3 this operation is likely time consuming this is acceptable as workaround
+      log       <- logTask
+      nextState <- setup.getNextState(keysR, currentState, keysChangedToken)
+      _         <- log.debug(s"Next state computed to be $nextState")
+      keys      <- keysR.get
+      optKey = setup.selectObjectForState(nextState, keys)
+      _ <- log.debug(s"Will ask for key $optKey") *> optKey
+        .map(key =>
+          zio.s3
+            .getObject(setup.bucketName, key)
+            .transduce(setup.transducer)
+            .foreach(value => q.offer(setup.deriveKafkaRecordKey(nextState, value) -> value))
+        )
+        .getOrElse(ZIO.fail(TamerError(s"File not found with key $optKey for state $nextState"))) // FIXME: relies on nextState.toString
+    } yield nextState).mapError(e => TamerError("Error while doing iterationTimeBased", e))
 }
```

Comment:
No more need for all of this since we can just block until the file list actually changed (in the other fiber and through notification in the `Queue[Unit]`.